api/test/e2enew/base/http.go

```diff
@@ -92,3 +104,44 @@ func BatchTestServerPort(times int) map[string]int {
 
 	return res
 }
+
+func GetReader(reqParams map[string]string, contentType string, files []UploadFile) (io.Reader, string, error) {
+	if strings.Index(contentType, "json") > -1 {
+		bytesData, _ := json.Marshal(reqParams)
+		return bytes.NewReader(bytesData), contentType, nil
+	}
+	if files != nil {
+		body := &bytes.Buffer{}
+		writer := multipart.NewWriter(body)
+		for _, uploadFile := range files {
+			file, err := os.Open(uploadFile.Filepath)
+			if err != nil {
+				return nil, "", err
+			}
+			part, err := writer.CreateFormFile(uploadFile.Name, filepath.Base(uploadFile.Filepath))
+			if err != nil {
+				return nil, "", err
+			}
+			_, err = io.Copy(part, file)
+			defer file.Close()
+		}
+		for k, v := range reqParams {
+			if err := writer.WriteField(k, v); err != nil {
+				return nil, "", err
+			}
+		}
+		if err := writer.Close(); err != nil {
+			return nil, "", err
+		}
+		return body, writer.FormDataContentType(), nil
+	}
+
+	urlValues := url.Values{}
+	for key, val := range reqParams {
+		urlValues.Set(key, val)
+	}
+
+	reqBody := urlValues.Encode()
+
+	return strings.NewReader(reqBody), contentType, nil
```

Comment:
Put the `param` to the body? This confused me.

Reply:
Yes, because the sending file is added, the data type in the body is bytes. If it is put in the body, it is not convenient to read. So I communicated with @liuxiran at that time, so it can only put other parameters in the header. Do you have a better suggestion?

Reply:
http.go In the getReader function, It from the old E2E/ http.go.

My idea is that E2E will be rewritten with the new ginkgo, so I migrate it directly here without modifying the code.

My understanding here is to deal with a variety of data formats, such as form-data, JSON, files.

The incoming params may be in a variety of situations. Here is the third one url.Values set reqParams map[string]string

Reply:
@nic-chen Do you know about this?

Reply:
I feel that there is a problem with this. The content of the file should not be read in the test case, but the content of the uploaded file should be read in the main program.

Reply:
@liuxiran 

Because the front-end has cross-domain problems, it needs to conduct a transparent transmission through the manage API. 

Here, the front-end places the file bytes in the body and transfers them to the back-end, and then directly transmits them through the back-end. Therefore, the E2E test needs to read the file here.

Reply:
Sorry, my question is that why we should put the `reqParams` to `reqBody`?

Reply:
Request params include url, prototal, method, header params and body params, they used to form the new request rewarding by golang.

Before this pr, we put all request params to the request body, and we use `json` to transfor data
due to this new feature, the origin request support not only `json`, `form-data` is accepted. Inorder not to disturb the origin formdata, we put others params including: url, prototal, method and header params to the request header, and the reqest body keep the origin body, so that the forwarding request can use the body directly.

Reply:
@liuxiran Thanks for the explanation. 😄